src/sqlfluff/parser/grammar.py

```diff
@@ -507,33 +517,54 @@ def _bracket_sensitive_look_ahead_match(cls, segments, matchers, parse_context,
                             # We've found an end bracket, remove it from the
                             # stack and carry on.
                             bracket_stack.pop()
+                            definite_stack.pop()
                             pre_seg_buff += pre
                             pre_seg_buff += match.matched_segments
                             seg_buff = match.unmatched_segments
                             continue
                         else:
                             raise RuntimeError("I don't know how we get here?!")
                     else:
-                        # No match and we're in a bracket stack. Raise an error
+                        # No match, we're in a bracket stack. Either this is an error,
+                        # OR we were mistaken in our initial identification of the opening
+                        # bracket. That's only allowed if `not definitely_bracket`.
+
+                        # Any any not definitely brackets
+                        if any(not definite for definite in definite_stack):
+                            # If so, remove the last one from the stack and try again.
+                            for idx, elem in enumerate(reversed(definite_stack)):
+                                if not elem:
+                                    del bracket_stack[-idx]
```

Comment:
This loop and the `del` lines are confusing. Maybe add a comment with an example or brief description of what this code is doing? Could the stack(s) be encapsulated in a class? Is there another way to code these `del` lines, e.g. something like `.pop()`? Could a class like [`deque`](https://docs.python.org/3/library/collections.html#collections.deque) potentially be useful here?